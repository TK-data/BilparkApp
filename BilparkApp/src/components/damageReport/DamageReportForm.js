import React, { Component } from 'react';
import t from 'tcomb-form-native';
import { connect } from 'react-redux';
import { NavigationActions } from 'react-navigation';
import { StyleSheet, Alert } from 'react-native';
import { Content, Button, Text, View, Spinner } from 'native-base';

import { postDamageReport, getCurrentDamageReport, damageReportOptions, damageReportValues } from '../../actions/damageReportForm';

const Items = [
  {
    ItemType: 'Wheel',
    Damaged: false,
  },
  {
    ItemType: 'Window',
    Damaged: false,
  },
  {
    ItemType: 'CarLight',
    Damaged: false,
  },
  {
    ItemType: 'FrontBumper',
    Damaged: false,
  },
  {
    ItemType: 'BackBumper',
    Damaged: false,
  },
  {
    ItemType: 'RightBodyWork',
    Damaged: false,
  },
  {
    ItemType: 'LeftBodyWork',
    Damaged: false,
  },
];

const styles = StyleSheet.create({
  form: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  },
  content: {
  },
  finnBil: {
    color: 'white',
    fontSize: 18,
    marginBottom: '8%',
  },
  button: {
    alignSelf: 'center',
    marginBottom: '3%',
  },
});

class DamageReportForm extends Component {
  componentDidMount() {
    this.props.getValues();
    this.props.changeOptions(this.props.values);
  }

  onChange(values) {
    this.props.changeOptions(values);
    Items[0].Damaged = values.FelgHjul;
    Items[0].Description = values.FelgHjulBeskrivelse;
    Items[1].Damaged = values.Glass;
    Items[1].Description = values.GlassBeskrivelse;
    Items[2].Damaged = values.LysUtvendig;
    Items[2].Description = values.LysUtvendigBeskrivelse;
    Items[3].Damaged = values.StøtfangerFront;
    Items[3].Description = values.StøtfangerFrontBeskrivelse;
    Items[4].Damaged = values.StøtfangerBak;
    Items[4].Description = values.StøtfangerBakBeskrivelse;
    Items[5].Damaged = values.KarosseriHøyre;
    Items[5].Description = values.KarosseriHøyreBeskrivelse;
    Items[6].Damaged = values.KarosseriVenstre;
    Items[6].Description = values.KarosseriVenstreBeskrivelse;
    this.props.updateValues(Items);
  }

  handleSubmit() {
    Items[0].Damaged = this.props.values.FelgHjul;
    Items[0].Description = this.props.values.FelgHjulBeskrivelse;
    Items[1].Damaged = this.props.values.Glass;
    Items[1].Description = this.props.values.GlassBeskrivelse;
    Items[2].Damaged = this.props.values.LysUtvendig;
    Items[2].Description = this.props.values.LysUtvendigBeskrivelse;
    Items[3].Damaged = this.props.values.StøtfangerFront;
    Items[3].Description = this.props.values.StøtfangerFrontBeskrivelse;
    Items[4].Damaged = this.props.values.StøtfangerBak;
    Items[4].Description = this.props.values.StøtfangerBakBeskrivelse;
    Items[5].Damaged = this.props.values.KarosseriHøyre;
    Items[5].Description = this.props.values.KarosseriHøyreBeskrivelse;
    Items[6].Damaged = this.props.values.KarosseriVenstre;
    Items[6].Description = this.props.values.KarosseriVenstreBeskrivelse;
    this.props.changeValues(Items);
  }

  render() {
    if (this.props.isLoading) {
      return (
        <View>
          <Spinner color="white" />
        </View>
      );
    }

    if (this.props.car === 'null') {
      return (
        <View>
          <Text style={styles.finnBil}>Du har ikke registrert en bil enda.. </Text>
          <Button
            bordered
            light
            onPress={() => { this.props.navigate('RegisterCar'); }}
            style={styles.button}
          >
            <Text>Finn din bil</Text>
          </Button>
        </View>
      );
    }

    const Damages = t.struct({
      KarosseriVenstre: t.Boolean,
      KarosseriVenstreBeskrivelse: t.maybe(t.String),
      KarosseriHøyre: t.Boolean,
      KarosseriHøyreBeskrivelse: t.maybe(t.String),
      StøtfangerFront: t.Boolean,
      StøtfangerFrontBeskrivelse: t.maybe(t.String),
      StøtfangerBak: t.Boolean,
      StøtfangerBakBeskrivelse: t.maybe(t.String),
      LysUtvendig: t.Boolean,
      LysUtvendigBeskrivelse: t.maybe(t.String),
      Glass: t.Boolean,
      GlassBeskrivelse: t.maybe(t.String),
      FelgHjul: t.Boolean,
      FelgHjulBeskrivelse: t.maybe(t.String),
    });
    const Form = t.form.Form;
    return (
      <View style={styles.form}>
        <Content style={styles.content}>
          <Form
            ref={c => this.form = c}
            type={Damages}
            options={this.props.formOptions}
            value={this.props.values}
            onChange={value => this.onChange(value)}
          />
        </Content>
        <Button
          bordered
          light
          style={styles.button}
          onPress={() => {
            Alert.alert(
              'Bekreft skademelding',
              'Er du sikker på at informasjonen er riktig?',
              [
                { text: 'Avbryt', style: 'cancel' },
                { text: 'Lagre', onPress: () => this.handleSubmit() },
              ],
            );
          }}
        >
          <Text> Registrer </Text>
        </Button>
        {this.props.success ? Alert.alert(
          '',
          'Skademeldingen ble lagret',
          [
            { text: 'OK' },
          ],
        ) : null }
      </View>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    isLoading: state.damageReportForm.isLoading,
    hasErrored: state.damageReportForm.hasErrored,
    currentDamageReport: state.damageReportForm.currentDamageReport,
    values: state.damageReportValues,
    car: state.auth.car,
    formOptions: state.damageReportOptions,
    success: state.damageReportForm.success,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    changeOptions: options => dispatch(damageReportOptions(options)),
    changeValues: ItemArray => dispatch(postDamageReport(ItemArray)),
    updateValues: values => dispatch(damageReportValues(values)),
    getValues: ItemArray => dispatch(getCurrentDamageReport(ItemArray)),
    navigate: (routeName) => {
      dispatch(NavigationActions.navigate({ routeName }));
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(DamageReportForm);
