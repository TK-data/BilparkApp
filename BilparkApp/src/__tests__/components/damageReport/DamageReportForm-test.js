import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import t from 'tcomb-form-native';
import { Button } from 'native-base';

import damageReportValues from '../../../actions/damageReportForm';

import DamageReportForm from '../../../components/damageReport/DamageReportForm';

const Items = [
  {
    ItemType: 'Wheel',
    Damaged: false,
  },
  {
    ItemType: 'Window',
    Damaged: false,
  },
  {
    ItemType: 'CarLight',
    Damaged: false,
  },
  {
    ItemType: 'FrontBumper',
    Damaged: false,
  },
  {
    ItemType: 'BackBumper',
    Damaged: false,
  },
  {
    ItemType: 'RightBodyWork',
    Damaged: false,
  },
  {
    ItemType: 'LeftBodyWork',
    Damaged: false,
  },
];

const initialState = {
  damageReportForm: {
    isLoading: true,
    hasErrored: false,
    currentDamageReport: [
      {
        ItemType: 'Wheel',
        Damaged: false,
      },
      {
        ItemType: 'Window',
        Damaged: false,
      },
      {
        ItemType: 'CarLight',
        Damaged: false,
      },
      {
        ItemType: 'FrontBumper',
        Damaged: false,
      },
      {
        ItemType: 'BackBumper',
        Damaged: false,
      },
      {
        ItemType: 'RightBodyWork',
        Damaged: false,
      },
      {
        ItemType: 'LeftBodyWork',
        Damaged: false,
      },
    ],
    values: [
      {
        ItemType: 'Wheel',
        Damaged: false,
      },
      {
        ItemType: 'Window',
        Damaged: false,
      },
      {
        ItemType: 'CarLight',
        Damaged: false,
      },
      {
        ItemType: 'FrontBumper',
        Damaged: false,
      },
      {
        ItemType: 'BackBumper',
        Damaged: false,
      },
      {
        ItemType: 'RightBodyWork',
        Damaged: false,
      },
      {
        ItemType: 'LeftBodyWork',
        Damaged: false,
      },
    ],
    testValues: false,
  },
  values: [
    {
      ItemType: 'Wheel',
      Damaged: false,
    },
    {
      ItemType: 'Window',
      Damaged: false,
    },
    {
      ItemType: 'CarLight',
      Damaged: false,
    },
    {
      ItemType: 'FrontBumper',
      Damaged: false,
    },
    {
      ItemType: 'BackBumper',
      Damaged: false,
    },
    {
      ItemType: 'RightBodyWork',
      Damaged: false,
    },
    {
      ItemType: 'LeftBodyWork',
      Damaged: false,
    },
  ],
  auth: {
    car: 'testcar',
  },
  damageReportOptions: {
    formOptions: 'testoptions',
  },
  currentDamageReport: {
    Items: 'testItems',
  },
};

const bigInitialState = {
  damageReportForm: {
    currentDamageReport: {
      CarID: 34,
      DamageReportID: 187,
      Items: [
        {
          DamageReportID: 187,
          Damaged: false,
          Description: null,
          ItemID: 92,
          ItemType: 'Wheel',
          createdAt: '2018-05-14T13:45:59.000Z',
          updatedAt: '2018-05-14T13:45:59.000Z',
        },
        {
          DamageReportID: 187,
          Damaged: true,
          Description: 'Bra nå',
          ItemID: 93,
          ItemType: 'BackBumper',
          createdAt: '2018-05-14T13:45:59.000Z',
          updatedAt: '2018-05-14T13:45:59.000Z',
        },
        {
          DamageReportID: 187,
          Damaged: false,
          Description: null,
          ItemID: 94,
          ItemType: 'Window',
          createdAt: '2018-05-14T13:45:59.000Z',
          updatedAt: '2018-05-14T13:45:59.000Z',
        },
        {
          DamageReportID: 187,
          Damaged: false,
          Description: null,
          ItemID: 95,
          ItemType: 'CarLight',
          createdAt: '2018-05-14T13:45:59.000Z',
          updatedAt: '2018-05-14T13:45:59.000Z',
        },
        {
          DamageReportID: 187,
          Damaged: true,
          Description: 'Est av emil',
          ItemID: 96,
          ItemType: 'FrontBumper',
          createdAt: '2018-05-14T13:45:59.000Z',
          updatedAt: '2018-05-14T13:45:59.000Z',
        },
        {
          DamageReportID: 187,
          Damaged: false,
          Description: null,
          ItemID: 97,
          ItemType: 'LeftBodyWork',
          createdAt: '2018-05-14T13:45:59.000Z',
          updatedAt: '2018-05-14T13:45:59.000Z',
        },
        {
          DamageReportID: 187,
          Damaged: false,
          Description: null,
          ItemID: 98,
          ItemType: 'RightBodyWork',
          createdAt: '2018-05-14T13:45:59.000Z',
          updatedAt: '2018-05-14T13:45:59.000Z',
        },
      ],
      UserID: 49,
      createdAt: '2018-05-14T13:45:59.000Z',
      updatedAt: '2018-05-14T13:45:59.000Z',
    },
    damageReports: [],
    hasErrored: false,
    isLoading: false,
  },
  damageReportOptions: {
    auto: 'placeholder',
    fields: {
      FelgHjul: {
        label: 'Hjul (felg)',
      },
      FelgHjulBeskrivelse: {
        hidden: true,
        placeholder: 'Beskrivelse av skaden (Valgfritt)',
      },
      Glass: {
        label: 'Vinduer',
      },
      GlassBeskrivelse: {
        hidden: true,
        placeholder: 'Beskrivelse av skaden (Valgfritt)',
      },
      KarosseriHøyre: {
        label: 'Høyre karosseri',
      },
      KarosseriHøyreBeskrivelse: {
        hidden: true,
        placeholder: 'Beskrivelse av skaden (Valgfritt)',
      },
      KarosseriVenstre: {
        label: 'Venstre karosseri',
      },
      KarosseriVenstreBeskrivelse: {
        hidden: true,
        placeholder: 'Beskrivelse av skaden (Valgfritt)',
      },
      LysUtvendig: {
        label: 'Lys (utvendig)',
      },
      LysUtvendigBeskrivelse: {
        hidden: true,
        placeholder: 'Beskrivelse av skaden (Valgfritt)',
      },
      StøtfangerBak: {
        label: 'Støtfanger Bak',
      },
      StøtfangerBakBeskrivelse: {
        hidden: true,
        placeholder: 'Beskrivelse av skaden (Valgfritt)',
      },
      StøtfangerFront: {
        label: 'Støtfanger front',
      },
      StøtfangerFrontBeskrivelse: {
        hidden: true,
        placeholder: 'Beskrivelse av skaden (Valgfritt)',
      },
    },
    stylesheet: {
      button: {
        alignSelf: 'stretch',
        backgroundColor: '#48BBEC',
        borderColor: '#48BBEC',
        borderRadius: 8,
        borderWidth: 1,
        height: 36,
        justifyContent: 'center',
        marginBottom: 10,
      },
      buttonText: {
        alignSelf: 'center',
        color: 'white',
        fontSize: 18,
      },
      checkbox: {
        error: {
          marginBottom: 4,
        },
        normal: {
          marginBottom: 20,
          marginTop: 10,
        },
      },
      controlLabel: {
        error: {
          color: '#fff',
          fontSize: 17,
          fontWeight: '500',
          marginBottom: 7,
        },
        normal: {
          color: '#fff',
          fontSize: 17,
          fontWeight: '500',
          marginBottom: 7,
        },
      },
      dateTouchable: {
        error: {},
        normal: {},
      },
      dateValue: {
        error: {
          color: '#a94442',
          fontSize: 17,
          marginBottom: 5,
          padding: 7,
        },
        normal: {
          color: '#000000',
          fontSize: 17,
          marginBottom: 5,
          padding: 7,
        },
      },
      datepicker: {
        error: {
          marginBottom: 4,
        },
        normal: {
          marginBottom: 4,
        },
      },
      errorBlock: {
        color: '#db2b1e',
        fontSize: 17,
        marginBottom: 2,
      },
      fieldset: {},
      formGroup: {
        error: {
          flexDirection: 'row',
          marginBottom: 10,
          width: 355,
        },
        normal: {
          alignItems: 'center',
          flexDirection: 'row',
          justifyContent: 'space-between',
          marginBottom: 10,
          paddingLeft: 10,
          paddingRight: 10,
          width: 355,
        },
      },
      helpBlock: {
        error: {
          color: '#999999',
          fontSize: 17,
          marginBottom: 2,
        },
        normal: {
          color: '#999999',
          fontSize: 17,
          marginBottom: 2,
        },
      },
      pickerContainer: {
        error: {
          borderColor: '#a94442',
          borderRadius: 4,
          borderWidth: 1,
          marginBottom: 4,
        },
        normal: {
          backgroundColor: '#fff',
          borderColor: '#cccccc',
          borderRadius: 4,
          borderWidth: 1,
          marginBottom: 4,
        },
        open: {},
      },
      pickerTouchable: {
        active: {
          borderBottomWidth: 1,
          borderColor: '#cccccc',
        },
        error: {
          alignItems: 'center',
          flexDirection: 'row',
          height: 44,
        },
        normal: {
          alignItems: 'center',
          flexDirection: 'row',
          height: 44,
        },
      },
      pickerValue: {
        error: {
          fontSize: 17,
          paddingLeft: 7,
        },
        normal: {
          fontSize: 17,
          marginLeft: 'auto',
          marginRight: 'auto',
          paddingLeft: 7,
        },
      },
      select: {
        error: {},
        normal: {},
      },
      textbox: {
        error: {
          backgroundColor: '#fff',
          borderColor: '#a94442',
          borderRadius: 4,
          borderWidth: 3,
          color: '#000000',
          fontSize: 20,
          height: 40,
          marginBottom: 5,
          paddingHorizontal: 7,
          paddingVertical: 7,
        },
        normal: {
          backgroundColor: '#fff',
          borderColor: '#cccccc',
          borderRadius: 4,
          borderWidth: 1,
          color: '#000',
          fontSize: 20,
          height: 40,
          marginBottom: 5,
          marginTop: -15,
          paddingHorizontal: 7,
          paddingVertical: 7,
          width: 337.5,
        },
        notEditable: {
          backgroundColor: '#eeeeee',
          borderColor: '#cccccc',
          borderRadius: 4,
          borderWidth: 1,
          color: '#777777',
          fontSize: 17,
          height: 36,
          marginBottom: 5,
          paddingHorizontal: 7,
          paddingVertical: 7,
        },
      },
      textboxView: {
        error: {},
        normal: {},
        notEditable: {},
      },
    },
  },
  damageReportValues: {
    FelgHjul: false,
    FelgHjulBeskrivelse: null,
    Glass: false,
    GlassBeskrivelse: null,
    KarosseriHøyre: false,
    KarosseriHøyreBeskrivelse: null,
    KarosseriVenstre: false,
    KarosseriVenstreBeskrivelse: null,
    LysUtvendig: false,
    LysUtvendigBeskrivelse: null,
    StøtfangerBak: true,
    StøtfangerBakBeskrivelse: 'Bra nå',
    StøtfangerFront: true,
    StøtfangerFrontBeskrivelse: 'Est av emil',
  },
  auth: {
    car: 'testcar',
  },
};


const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe('Testing DamageReportForm', () => {
  let store;
  // set up a fake store for all our tests
  beforeEach(() => {
    store = mockStore({ values: {} });
  });

  it('Renders as expected', () => {
    const comp = shallow(
      <DamageReportForm />,
      { context: { store: mockStore(initialState) } },
    );
    expect(comp.dive()).toMatchSnapshot();
  });
  const wrapper = shallow(
    <DamageReportForm value={Items} />,
    { context: { store: mockStore(bigInitialState) } },
  );
  const Form = t.form.Form;
  it('Containts the form', () => {
    const render = wrapper.dive().dive();
    expect(render.find(Form).exists());
  });
  it('Contains the register button', () => {
    const render = wrapper.dive().dive();
    expect(render.find(Button).exists());
  });
  it('Can press the register button', () => {
    // const render = wrapper.dive().dive();
    // const registerButton = render.find(Button);
    // registerButton.simulate('press');
  });
  it('Can use the onChange function', () => {
    const values = {
      FelgHjul: false,
      FelgHjulBeskrivelse: null,
      Glass: false,
      GlassBeskrivelse: null,
      KarosseriHøyre: false,
      KarosseriHøyreBeskrivelse: null,
      KarosseriVenstre: false,
      KarosseriVenstreBeskrivelse: null,
      LysUtvendig: false,
      LysUtvendigBeskrivelse: null,
      StøtfangerBak: false,
      StøtfangerBakBeskrivelse: '',
      StøtfangerFront: false,
      StøtfangerFrontBeskrivelse: '',
    };
    wrapper.dive().instance().onChange(values);
  });
  it('Can use the handleSubmit function', () => {
    wrapper.dive().instance().handleSubmit();
  });
});
