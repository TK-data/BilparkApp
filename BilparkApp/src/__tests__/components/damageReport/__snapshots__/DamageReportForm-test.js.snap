// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing DamageReportForm Renders as expected 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DamageReportForm
    car="testcar"
    changeOptions={[Function]}
    changeValues={[Function]}
    currentDamageReport={
      Array [
        Object {
          "Damaged": false,
          "ItemType": "Wheel",
        },
        Object {
          "Damaged": false,
          "ItemType": "Window",
        },
        Object {
          "Damaged": false,
          "ItemType": "CarLight",
        },
        Object {
          "Damaged": false,
          "ItemType": "FrontBumper",
        },
        Object {
          "Damaged": false,
          "ItemType": "BackBumper",
        },
        Object {
          "Damaged": false,
          "ItemType": "RightBodyWork",
        },
        Object {
          "Damaged": false,
          "ItemType": "LeftBodyWork",
        },
      ]
    }
    formOptions={
      Object {
        "formOptions": "testoptions",
      }
    }
    getValues={[Function]}
    hasErrored={false}
    isLoading={true}
    navigate={[Function]}
    success={undefined}
    updateValues={[Function]}
    values={undefined}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Styled(Spinner)
        color="white"
        virtual={undefined}
      />,
      "virtual": undefined,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "color": "white",
        "virtual": undefined,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Styled(Spinner)
          color="white"
          virtual={undefined}
        />,
        "virtual": undefined,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "color": "white",
          "virtual": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
    },
  },
}
`;
